### PR TITLE
build: bundle recast

### PR DESCRIPTION
**Pull request by antfu**
Install size from 8.5MB -> 5.2MB (mainly Babel), from 35 packages to 4 packages. Get rid of many es5 polyfills and `ESM -> CJS -> ESM` glues. Result in around 5% faster, tho the performance isn't very related to our use cases.

### CODE REVIEW

**Comment by pi0** [comment]
It is a nice idea btw have you tried with impact on nuxi final size with double bundling?

might worth to have a subpath with optional non bundled version.

File: package.json [hunk 1]
```diff
@@ -27,34 +27,47 @@
   ],
   "scripts": {
     "build": "unbuild",
+    "prepare": "esno ./scripts/vendor.ts",
```
Comment (pi0) [inline package.json:30]: Why not jiti 😋

**Comment by antfu** [comment]
Babel is not bundled. Only `recast` and `ast-types` are bundled from the source code, as their distribution targets Node 4 (es2015 and transpiled to CJS). This approach allows us to ship ESM directly without `ESM -> CJS -> ESM` convention.

File: scripts/vendor.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+// This script clones recast and patches, and then re-bundle it so we get rid of the unnecessary polyfills
+
+async function cloneRecast() {
+  if (fs.existsSync('vendor/recast')) {
+    console.log('vendor/recast already exists')
+  }
+  else {
+    // Clone recast
+    await execa('npx', ['tiged', 'benjamn/recast#v0.23.4', 'vendor/recast'], { stdio: 'inherit' })
```
Comment (pi0) [inline scripts/vendor.ts:14]: Why not giget? :))

File: package.json (hunk 1 above)
Comment (antfu) [reply package.json:30]: I would like to have top-level await landed in jiti 👀

File: scripts/vendor.ts (hunk 2 above)
Comment (antfu) [reply scripts/vendor.ts:14]: 👀 I didn't knew it :P

File: src/helpers/vite.ts [hunk 3]
```diff
@@ -133,7 +133,7 @@ function insertPluginIntoVariableDeclarationConfig(
     ) {
       // @ts-ignore this works despite the type error because of recast
       declaration.init = generateCode(
-        builders.functionCall(declaration.init.callee.name, configObject)
+        builders.functionCall(declaration.init.callee.name, configObject),
```
Comment (pi0) [inline src/helpers/vite.ts:136]: Maybe lint in main to reduce diff?

File: package.json (hunk 1 above)
Comment (pi0) [reply package.json:30]: Let’s make that happen!

File: package.json [hunk 4]
```diff
@@ -27,34 +27,47 @@
...
-    "test": "vitest"
+    "test": "vitest",
+    "test:build": "TEST_BUILD=true vitest",
+    "test:full": "pnpm run test && pnpm run build && pnpm run test:build"
   },
   "dependencies": {
     "@babel/parser": "^7.22.15",
     "@babel/types": "^7.22.15",
-    "recast": "^0.23.4"
+    "source-map-js": "^1.0.2"
```
Comment (pi0) [inline package.json:45]: Shall we somehow mock and remove this too or it is essential dep?

Comment (antfu) [reply package.json:45]: We do expose an `map` object, but I am not sure if anyone is using it. If you think that makes sense, we could introduce a breaking change and not to expose it any longer.

Comment (pi0) [reply package.json:45]: A major version to allow saving as much as possible seems a nice idea! (BTW i will try to locally have a look before release if we can save more, it is a nice perf PR!)

**Comment by pi0** [comment]
Small notes:

`assert: resolve(__dirname, 'stubs/assert/index.js'),` this from `build.config` seems unused


Enabling `rollup.esbuild.minify: true` reduces bundle size from 640kB to 380kB!

**Comment by antfu** [comment]
Screenshot 2023-09-15 at 10 17 13

After running `pnpm pack`, the size doesn't have that much different with gzip. I would think it would be better trade-off to avoid minification for better DX and debugging experience.

**Comment by pi0** [comment]
since we are not bundling babel stuff yet, yes it is not much different. anyway consider that when bundling more, the startup (and sometimes runtime) time also improves with modification + memory usage is always less.

**Comment by antfu** [comment]
Yeah, but I imagine we won't bundle Babel any sooner as it's quite big and commonly used, isn't it? I imagine we can bundle recast because it's a bit niche, and probably only used by `magicast`.

**Comment by pi0** [comment]
Sure. That's what I wanted to share with you later about babel ;)

**Pull request merged**